### PR TITLE
[DRAFT] Biome API 1.16.2

### PR DESCRIPTION
**Pull request by CheaterCodes**
Because of too many changes to world generation, the old biome API requires a complete rewrite.

This is my initial attempt at a 1.16.2 biome API.
While I'm confident in most of my code, if this *does* get merged, it should probably still be merged as an experimental feature since the underlying vanilla code is also experimental.

While this is a draft, I'd like to get this out reasonably fast, even if incomplete as a experimental API since many mods need it.
Also please help me with this through discussions or critique, I am pretty new to fabric api development.

**This requires #1029** (will rebase once that gets merged)
**This replaces #982**
**This replaces #996**

## Todo List

- [x] Register new Biomes
- [ ] Add Overworld Biomes to vanilla world gen
- [x] Add Nether Biomes to vanilla world gen
- [ ] Add End Biomes to vanilla world gen (Probably not happening)
- [x] Modify existing Biomes (Covers #982 )
- [ ] Easier addition of features to the builder (e.g. using RegistryKey instead of a supplier)
- [ ] Provide access to the DynamicRegistryManager in the callback (also needs to be changed in #1029 )

## About Overworld Biomes
I don't actually know how feasible it is to provide a proper API for this.
It seems like the mess in `BiomeLayer.build` is responsible for selecting the biomes to generate and it seems to use a lot of hardcoded biome ids.
Please let any ideas be known.

### CODE REVIEW

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java
Comment (i509VCB) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java:6]: The rename really isn't necessary tbf

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java
Comment (i509VCB) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java:12]: What on earth

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/event/BiomeLoadingCallback.java
Comment (i509VCB) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/event/BiomeLoadingCallback.java:20]: This would definitely let you add things to biomes. My concern is mod devs use this to change attributes about a biome.

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/FabricBiomesEntrypoint.java
Comment (i509VCB) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/FabricBiomesEntrypoint.java]: This entire PR is a really hard sell for me if it depends on another PR that has not been merged yet.

~~I don't see any real reason to review until the backing logic is merged.~~

~~Gonna have to stop my review here.~~

Edit: I probably should review public api here tbf.

**Review by i509VCB — CHANGES_REQUESTED** [review]

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java
Comment (CheaterCodes) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java:6]: This isn't a rename, this is a complete rewrite.
The old api got completely removed.
That being said, if the older name is better I can change it to that

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/event/BiomeLoadingCallback.java
Comment (CheaterCodes) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/event/BiomeLoadingCallback.java:20]: I mean, if mod devs want to change attributes, they might as well?

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/FabricBiomesEntrypoint.java
Comment (CheaterCodes) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/FabricBiomesEntrypoint.java]: That's unfortunate. But this is a draft after all, so I don't expect full reviews.
I'm just hoping on some input.

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java
Comment (CheaterCodes) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java:12]: What on earth what?

Comment (i509VCB) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java:12]: It's definitely a bit odd to have a public api method taking a biome registry ref and biome. To me it seems like an internal map of `RegistryKey<Biome>` to `Biome`

Comment (CheaterCodes) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java:12]: So you'd expect an identifier instead?
I tried to mimic the `Biomes` class here, which does it similarly.
Actually, `BuiltinRegistries.BIOME` also does it like this.

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/event/BiomeLoadingCallback.java
Comment (CheaterCodes) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/event/BiomeLoadingCallback.java:20]: I could reduce the scope of the builder to only features, but I don't see how that would make it better.
This builder can also be used to clone existing biomes for re-registering them as something else, in that case you'd be happy to be able to change those attributes.

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java
Comment (i509VCB) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java:12]: I had no idea what it did and it seems like the old equivalent of regular register method.

Probably fine

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java
Comment (i509VCB) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java]: Hmm, since we broke api, maybe return `RegistryKey<Biome>` in these methods so we could chain these calls like a

```java
public static final RegistryKey<Biome> TEST_PLAINS = addToOverworld(FabricBiomes.register(RegistryKey.of(Registry.BIOME_KEY, new Identifier(MOD_ID, "test_plains"), DefaultBiomeCreator.createPlains(false)));
```

Yes it's long but it can allow for mod devs to make some helper methods in their registry classes.

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java
Comment (CheaterCodes) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java:12]: I could also just expose it as `FabricBiomes.REGISTRY` and then expect people to do:
`Registry.register(FabricBiomes.REGISTRY, identifier, biome`.
I was kind of hoping though, that doing it this way I could discourage people saving biome instances, which can cause issues in 1.16.2.

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java
Comment (CheaterCodes) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java]: done 👍

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java
Comment (liach) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java:35]: How compatible is this with JSON biomes?

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java
Comment (liach) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java:12]: Just take identifier. Taking registry key is like the worst option

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/event/BiomeLoadingCallback.java
Comment (liach) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/event/BiomeLoadingCallback.java:20]: Changing attributes of JSON defined biomes is really concerning.

**Review by liach — CHANGES_REQUESTED** [review]
In all, I quite disagree with your design philosophy. I suggest moving all these processing as part of the Codec that handles the biome (the map codec) rather than injecting mid way

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java
Comment (shartte) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java:35]: JSON biomes are deserialized into the same class as built-in biomes, so there is no difference at this point anymore, since this class is constructed from a built Biome instance.

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java
Comment (shartte) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java:12]: All Vanilla internals now deal with Registry Keys. Anyone working with biomes will probably need to keep a RegistryKey for their Biome around.
If you want to add _another_ Overload. Fine.

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/event/BiomeLoadingCallback.java
Comment (i509VCB) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/event/BiomeLoadingCallback.java:20]: I had one solution around this, split the builder in two and the full builders implements the structure et al stuff from scoped builder

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java
Comment (i509VCB) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java]: Maps can be final

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java
Comment (i509VCB) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java]: Style violation. Space between variable and for loop.

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java
Comment (i509VCB) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java:15]: All the javadoc here went the way of the dodo.

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java
Comment (i509VCB) [inline fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java:143]: None of the parameters here are validated

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/event/BiomeLoadingCallback.java
Comment (shartte) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/event/BiomeLoadingCallback.java:20]: I wonder though: Why is changing structures/features different from changing other attributes?
If this is strictly for adding features and structures, we can probably reduce the surface of this API.

**Comment by shartte** [comment]
> In all, I quite disagree with your design philosophy. I suggest moving all these processing as part of the Codec that handles the biome (the map codec) rather than injecting mid way

Moving this to the Codec is insufficient. The codec has no knowledge of the ID of the Biome being deserialized, making it impossible to adapt the customization based on tags/whitelists/blacklist/etc.

**Comment by liach** [comment]
Why would you edit anything based on an arbitrary ID? This will make a feature exclusive to one biome while other custom biomes have no access to it. Moreover, that exclusive stuff you add with code might not work when a data pack intend to override everything in a hardcoded biome. In this case it will break hard.

**Comment by shartte** [comment]
@liach Mods have had configuration files with Biome black- and whitelists for their Worldgen for a long time.

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java
Comment (CheaterCodes) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java:35]: Yes, they behave the same.

**Comment by liach** [comment]
hmm, we can maybe make the biome from codec be aware of an id to setting function that allow polling custom settings by id. also biome has a codec beyond a map codec. think that codec has some knowledge about id

**Comment by CheaterCodes** [comment]
If I want to add a structure to the plains biome, I need to have an identifier to check if it's a plains biome.
The codec provides no way of knowing which biome your adding features into.
If I want to add to all biomes, I can still do it.
Trying to inject into codec just to inject into codec seems wrong.

**Comment by liach** [comment]
preferably you should move stuff to codec as much as possible still if they don't need id

**Comment by CheaterCodes** [comment]
I already chose this implementation over making the biomes mutable because that didn't feel right.
Ideally I'd love to intercpt the biomes while they're being created, but I'm not sure that's feasible.
If you find a nice injection point for that, let me know.

**Comment by CheaterCodes** [comment]
Everything needs an ID if I want a certain feature to only affect certain biomes.
And even if it didn't, I don't think it makes sense to split the implementation into two points.

**Comment by shartte** [comment]
@liach But why though? This doesn't actually change the public facing API as far as I could tell, and if this injection point works, why go yet another round at trying to find a different injection point without clear benefits? (Unless the benefits are just not clear _to me_).
p.s.: I have yet to actually dig deeper into the actual mixin/impl code of this PR. So take my opinion on this with a grain of salt. My initial idea was just append a "map" to the CODEC, but that fails because of DFU-insanity and id-loss.

**Comment by CheaterCodes** [comment]
The impl of this is actually really simple: I just attach a registry event to the dynamic registries.
So the modification part is rather straightforward, it just depends on #1029 .
Attaching the biomes to overworld world gen is my biggest concern and apparently done (to some degree) in #996 .
I'll need to take another look at that pr.

**Comment by liach** [comment]
The ideal situation is to move those thing you want to add to a biome to the biome definition itself than something externally associated with an ID. Like you can add a flag like "overworld biome". In the codec you scan if the flag is present; if it's absent you then compute a default value from other traits of this biome. The ID should be never used as a reliable trait; in worst case where you have to use an ID, you can consider an ID to object funtion (like how advancements look up parents)

**Comment by CheaterCodes** [comment]
Assume you want to add a feature to a single biome (like I did in the testmod) , how would you go about it?
> you can consider an ID to object funtion

That's exactly what a registry is?
If you know of a better way, let me know, but this is also analog to the loot table modification fapi offers.

**Comment by liach** [comment]
yes, so when you construct the biome you just pass a "registry -> element" function as part of the biome and pass in the active biome registry when the biome's function is used

**Review by valoeghese — CHANGES_REQUESTED** [review]
Overall seems promising. Using `RegistryKey<Biome>` is definitely important for how biomes work now, as it means we also support using plain json biomes if a modder wishes to use them.

However, I notice you don't touch the biome layers. The List of biomes in the overworld biome source is for "allowed biomes" in computation of the available structures and surfaces. To actually add such a thing to world generation, you need to modify biome layers.

Refer to the v1 biome api for the layer modifications for hills, shores, variants, etc. It should be almost identical, except you're dealing with registry keys. Ignore the fact that vanilla hardcodes the int ids in their part of biome layers, as our layer code should still be fine.

**Comment by valoeghese** [comment]
Preferably the OverworldBiomes API should look like the v1 biome api for easy porting, but take a `RegistryKey<Biome>` to replace the original `Biome` parameter.

**Comment by CheaterCodes** [comment]
Yes, the layers scared me a bit at first which is why they didn't make it into this draft PR.
I only later realized that this was already done in the previous api. 
I think it could be do better but I don't know how yet. 
Thanks for the feedback!

**Comment by Yoghurt4C** [comment]
:ant: draft it up boy :ant:

File: fabric-biomes-v1/build.gradle
Comment (i509VCB) [inline fabric-biomes-v1/build.gradle:1]: Since the old module is effectively dead, should there be change name to meet new naming standards.

Really a question for modmuss and player

**Comment by NeusFear** [comment]
I agree with the things about layers needing re-added. Apart from that and the style issues, this looks very promising! I'll be checking back often to see how it progresses.

**Comment by TelepathicGrunt** [comment]
Agreed Neus. Having layers being added too will help a lot of mods with how to do biome IDs in layers properly for their custom biome sources too. And I absolutely want the registry keys/IDs too as I check those for if I am adding/removing boulders from certain vanilla Giant Tree Taiga biomes. And I do check biome identifiers for the word "crimson" to know if I should add a crimson variant of a structure to it (cannot do translated name as that's client sided but once someone puts together biome tags, maybe I'll revisit this). But yeah, IDs please. No codecs

File: fabric-biomes-v1/build.gradle
Comment (modmuss50) [reply fabric-biomes-v1/build.gradle:1]: Yes, no reason not to.

**Comment by bluebear94** [comment]
On a sidenote, I wonder how viable a JSON-patching mod such as (shameless self-plug) [hyphae-fabric](https://gitlab.com/Kyarei/hyphae-fabric) would be at some of the same problems that this PR is addressing. I've written this mod to address problems around data packs having to replace things like loot tables or worldgen files wholesale and thus not being able to interoperate, though in the case of biomes, load order becomes a problem (e.g. any patches to biome JSONs should fire after all data packs that add biomes).

File: fabric-biomes-v1/build.gradle
Comment (CheaterCodes) [reply fabric-biomes-v1/build.gradle:1]: So it would be fabric-biome-api-v1 then?

**Comment by CheaterCodes** [comment]
Thanks for all the feedback!
Yes, the layer modification needs to return in some way or another. I'll work on it once I have the time. 

Json-Patching helps *somewhat*. I think hyphae-fabric looks really good. I'd have to look when and where it patches, but something like TelepathicGrunt mentioned (checking for "crimson" in identifiers) wouldn't be covered by that mod.

**Comment by shartte** [comment]
@bluebear94 Vanilla Biomes are not defined in JSON files.

**Comment by bluebear94** [comment]
Darn, I got my hopes up about that.

**Comment by ghost** [comment]
Personally, I'd really appreciate it if while we we're writing out this api version, we solidified a series of interfaces for a fabric biome, fabric feature, etc... I could very well see a situation in the near future where Mojang decides to completely switch to a JSON based biome system in which case they'd likely delete the current Biome and Feature classes.

The way I see it biome generation is kind of more like art, so every time you have to switch the medium a biome is expressed in, a little bit of the characteristics of the biome are lost in translation leading to generation disconnects. By having a clear and guaranteed interface of what a biome looks like in code, it would give a nice assurance that the "art" part of biome generation won't have to change and we can simply swap out the backend of how that biome interfaces with the rest of the game (whether that's mixing into the generators, or using the objects to generate JSON files).

Basically I'm just asking that we clearly define (through a series of interfaces) what a code based biome looks like so that version updates hopefully never have to entail completely rewriting biomes or features from scratch (as long as they implement the interfaces), instead we can just swap out the backend whenever Mojang changes something. Personally, that would just give me some good peace of mind. I don't know if that's on the table though.

**Comment by ghost** [comment]
I don't know if this helps, but for overworld it sounds like preliminary users have had success with applying an access widener to the arrays in `SetBiomesBaseLayer` and adding the raw ids in the biome registry to those arrays in order to generate the biomes. With fapi doing this it should be possible to simplify it to one overwrite instead of most people have been doing which is integrating, reconstructing, and overwriting the arrays whenever another biome is added. Also it might be possible to get rid of the access widener, but that's above my knowledge you'll have to look at it.

File: fabric-biomes-v1/build.gradle
Comment (i509VCB) [reply fabric-biomes-v1/build.gradle:1]: yes

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java
Comment (CheaterCodes) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java]: done

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java
Comment (CheaterCodes) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java]: fixed all style

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java
Comment (CheaterCodes) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomes.java:15]: putting this off until the api seems solid

File: fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java
Comment (CheaterCodes) [reply fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/FabricBiomeBuilder.java:143]: putting this off until the api seems solid

**Comment by MarcusElg** [comment]
Is there an ETA for this?

**Comment by CheaterCodes** [comment]
Basically look at #1053 for a straightup port of the old api that could possibly be merged soon.
I'm currently not really able to do much for this, but I'll come back at it at some point.

**Comment by i509VCB** [comment]
Closing this since #1053 and #1097 cover this whole pr

**Pull request closed without merging**
